app/addons/index.js

```diff
@@ -1,78 +1,339 @@
+    });
+  }
+
+  static _dependencyMet(name, dependency) {
+    try {
+      require.resolve(dependency);
+      return true;
+    } catch (e) {
+      logger.warn(`Npm package(name: ${dependency}) is not found, required by addon(name: ${name})`);
+      return false;
+    }
+  }
+
+  static _requirePackageFile(pName, pAddonPackageFilePath) {
+    const promiseToRequire = (new Promise((resolve) => {
+      const packageJson = require(addonPackageFilePath); // eslint-disable-line
+      resolve(packageJson);
+    })).catch((pError) => {
+      const error = 'Could not require addon package.json';
+      const meta = [`Filename: ${pName}`, `Filepath: ${pAddonPackageFilePath}`, `Error   : ${pError.message}`.red];
+      const solution = 'ensure that path exists and has no syntax errors';
+      return Promise.reject(new Error(createErrorMessage(error, meta, solution)));
+    });
+    return promiseToRequire;
+  }
+
+  static _requireModule(pName, pModulePath) {
+    const promiseToRequire = (new Promise((resolve) => {
+      const AddonModule = require(modulePath); // eslint-disable-line
+      resolve(AddonModule);
+    })).catch((pError) => {
+      const relativeAddonPath = path.relative('.', pModulePath);
+      const error = 'Could not require addon';
+      const meta = [`Filename: ${pName}`, `Filepath: ${relativeAddonPath}`, `Error   : ${pError.message}`.red];
+      const solution = 'ensure that path exists and module has no syntax errors';
+      return Promise.reject(new Error(createErrorMessage(error, meta, solution)));
+    });
+    return promiseToRequire;
+  }
+}
+
+class AddonManager {
+  constructor(app, server, io) {
+    this.app = app;
+    this.server = server;
+    this.io = io;
+
+    this.addons = [];
+    app.get('/addons', this.listAddons.bind(this));
+  }
+
+  get availableModules() {
+    return this.addons.map((addon) => {
+      const obj = { name: addon.name, state: 'active' };
+      return obj;
+    });
+  }
+
+  loadAddons() {
+    logger.info('Loading addons...');
+    logger.debug(`used directory: ${__dirname}`);
+
+    // Function that returns whether a file is a directory or not
+    function isAddon(file) {
+      return fs.lstatSync(path.join(__dirname, file)).isDirectory();
+    }
+
+    // Iterate through all directory files in the addons folder
+    return Promise.all(fs.readdirSync(__dirname).filter(isAddon).map((file) => {
+      const newAddon = new Addon(file);
+      this.addons.push(newAddon);
+      return newAddon.loadModule(file).catch((pError) => {
+        logger.warn(pError.message);
+        logger.debug(pError.stack);
+      });
+    }));
+  }
+
+  registerAddons() {
+    logger.info('Registering addons...');
+
+    // Promise to register all addons
+    const registerPromises = this.addons.filter((addon) => {
+      return addon.status.state === STATES.load && addon.status.phase === PHASES.done;
+    }).map((addon) => {
+      addon.createInstance(this.app, this.server, this.io);
```

Comment:
if one create instance fails is it halting rest of plugin registering... ? looks like it does.. maybe it should still try rest of plugins..